lib/python/treadmill_aws/garbage_collector.py

```diff
@@ -0,0 +1,96 @@
+"""Garbage collector for LDAP and IPA.
+"""
+
+import logging
+from os import path
+from os import environ as envvar
+from datetime import datetime
+from datetime import timedelta
+from functools import lru_cache
+from urllib3.exceptions import SSLError
+
+import jmespath
+from treadmill import admin
+from treadmill import context
+from treadmill_aws.awscontext import GLOBAL
+
+_LOGGER = logging.getLogger(__name__)
+
+_INSTANCE_NAME = jmespath.compile(
+    "Reservations[].Instances[].[Tags[?Key=='Name'].Value][][]")
+
+
+@lru_cache(maxsize=None)
+def _valid_ec2_instances():
+    _LOGGER.info('Fetching valid instances from AWS')
+    instances = _INSTANCE_NAME.search(
+        GLOBAL.ec2.describe_instances(
+            Filters=[{"Name": "instance-state-name",
+                      "Values": ["running", "pending", "rebooting",
+                                 "stopping", "stopped"]}]))
+    _LOGGER.debug("%d valid instances found", len(instances))
+
+    return set(instances)
+
+
+def _ipa_client():
+    # if 'IPA_DOMAIN' in envvar:
+    #     GLOBAL.ipa_domain = envvar["IPA_DOMAIN"]
+    # else:
+    #     _LOGGER.warning('IPA_DOMAIN environment variable is unset')
+    #
+    if envvar.get("IPA_CACERT") and path.exists(envvar.get("IPA_CACERT")):
```

Comment:
this is not needed. look at treadmill_aws/cli/admin/aws/__init__.py - it takes care of ipa certs. generally speaking we do not query environment in the inner modules, rely on parameters passsed.

awscontext.GLOBAL.ipaclient is all you need. look it up in code.

Reply:
fixed by moving back garbage_collector from core to treadmill_aws